journal/journal.go

```diff
@@ -0,0 +1,211 @@
+	SD_JOURNAL_FIELD_GID          = "_GID"
+	SD_JOURNAL_FIELD_HOSTNAME     = "_HOSTNAME"
+	SD_JOURNAL_FIELD_MACHINE_ID   = "_MACHINE_ID"
+)
+
+// Journal event constants
+const (
+	SD_JOURNAL_NOP        = int(C.SD_JOURNAL_NOP)
+	SD_JOURNAL_APPEND     = int(C.SD_JOURNAL_APPEND)
+	SD_JOURNAL_INVALIDATE = int(C.SD_JOURNAL_INVALIDATE)
+)
+
+// A Journal is a Go wrapper of an sd_journal structure.
+type Journal struct {
+	cjournal *C.sd_journal
+	mu       sync.Mutex
+}
+
+// A Match is a convenience wrapper to describe filters supplied to AddMatch.
+type Match struct {
+	Field string
+	Value string
+}
+
+// String returns a string representation of a Match suitable for use with AddMatch.
+func (m *Match) String() string {
+	return m.Field + "=" + m.Value
+}
+
+// NewJournal returns a new Journal instance pointing to the local journal
+func NewJournal() (*Journal, error) {
+	j := &Journal{}
+	err := C.sd_journal_open(&j.cjournal, C.SD_JOURNAL_LOCAL_ONLY)
+
+	if err < 0 {
+		return nil, fmt.Errorf("failed to open journal: %s", err)
+	}
+
+	return j, nil
+}
+
+func (j *Journal) Close() error {
+	j.mu.Lock()
+	C.sd_journal_close(j.cjournal)
+	j.mu.Unlock()
+	return nil
+}
+
+func (j *Journal) AddMatch(match string) error {
+	m := C.CString(match)
+	defer C.free(unsafe.Pointer(m))
+
+	j.mu.Lock()
+	C.sd_journal_add_match(j.cjournal, unsafe.Pointer(m), C.size_t(len(match)))
+	j.mu.Unlock()
+
+	return nil
+}
+
+func (j *Journal) Next() (int, error) {
+	j.mu.Lock()
+	r := C.sd_journal_next(j.cjournal)
+	j.mu.Unlock()
+
+	if r < 0 {
+		return int(r), fmt.Errorf("failed to iterate journal: %d", r)
+	}
+
+	return int(r), nil
+}
+
+func (j *Journal) Previous() (uint64, error) {
+	j.mu.Lock()
+	r := C.sd_journal_previous(j.cjournal)
+	j.mu.Unlock()
+
+	if r < 0 {
+		return uint64(r), fmt.Errorf("failed to iterate journal: %d", r)
+	}
+
+	return uint64(r), nil
+}
+
+func (j *Journal) PreviousSkip(skip uint64) (uint64, error) {
+	j.mu.Lock()
+	r := C.sd_journal_previous_skip(j.cjournal, C.uint64_t(skip))
+	j.mu.Unlock()
+
+	if r < 0 {
+		return uint64(r), fmt.Errorf("failed to iterate journal: %d", r)
+	}
+
+	return uint64(r), nil
+}
+
+func (j *Journal) GetData(field string) (string, error) {
+	f := C.CString(field)
+	defer C.free(unsafe.Pointer(f))
+
+	var d unsafe.Pointer
+	var l C.size_t
+
+	j.mu.Lock()
+	err := C.sd_journal_get_data(j.cjournal, f, &d, &l)
+	j.mu.Unlock()
+
+	if err < 0 {
+		return "", fmt.Errorf("failed to read message: %d", err)
+	}
+
+	msg := C.GoStringN((*C.char)(d), C.int(l))
+	return msg, nil
+}
+
+func (j *Journal) GetRealtimeUsec() (uint64, error) {
+	var usec C.uint64_t
+
+	j.mu.Lock()
+	r := C.sd_journal_get_realtime_usec(j.cjournal, &usec)
+	j.mu.Unlock()
+
+	if r < 0 {
+		return 0, fmt.Errorf("error getting timestamp for entry: %d", r)
+	}
+
+	return uint64(usec), nil
+}
+
+func (j *Journal) Print(priority int, message string) error {
+	m := C.CString(message)
+	defer C.free(unsafe.Pointer(m))
+
+	j.mu.Lock()
```

Comment:
Ehhhhh. I think we should follow Zbigniew's comment @ http://lists.freedesktop.org/archives/systemd-devel/2014-June/020067.html and split out write/read, there's really no need for Print to have access to the journal object here at all (and hence the lock is pointless)